lib/date.js

```diff
@@ -36,34 +45,56 @@ Hoek.inherits(internals.Date, Any);
 
 internals.Date.prototype._base = function (value, state, options) {
 
+    if (this._unit) {
+        internals.isValidUnit(this._unit);
+    }
+
     const result = {
-        value: (options.convert && internals.toDate(value, this._flags.format)) || value
+        value: (options.convert && internals.toDate(value, this)) || value
     };
 
     if (result.value instanceof Date && !isNaN(result.value.getTime())) {
         result.errors = null;
     }
     else {
-        result.errors = Errors.create(internals.isIsoDate(this._flags.format) ? 'date.isoDate' : 'date.base', null, state, options);
+        let type;
+        if (this._flags.unix) {
+            type = 'unix';
+        }
+        else if (this._flags.timestamp) {
+            type = 'timestamp';
+        }
+        else if (internals.isIsoDate(this._flags.format)) {
+            type = 'isoDate';
+        }
+        else {
+            type = 'base';
+        }
+
+        result.errors = Errors.create('date.' + type, null, state, options);
     }
 
     return result;
 };
 
 
-internals.toDate = function (value, format) {
+internals.toDate = function (value, obj) {
+
+    obj = obj || { _flags: {} };
+    const format = obj._flags.format;
+    const multiplier = obj._unit === 'seconds' ? 1000 : null;
 
     if (value instanceof Date) {
         return value;
     }
 
     if (typeof value === 'string' ||
-        Hoek.isInteger(value)) {
+        (typeof value === 'number' && !isNaN(value) && isFinite(value))) {
 
         if (typeof value === 'string' &&
-            /^[+-]?\d+$/.test(value)) {
+            /^[+-]?\d+(\.\d+)?$/.test(value)) {
 
-            value = parseInt(value, 10);
+            value = Number(value);
```

Comment:
I changed this to the Number constructor because `parseInt` drops the decimals. After [reviewing performance](https://jsperf.com/number-vs-parseint-vs-plus) I've updated it to use `parseFloat`.